tower-http/src/trace.rs

```diff
@@ -0,0 +1,230 @@
+        let start = Instant::now();
+
+        let method = req.method();
+        let path = if self.record_full_uri {
+            req.uri().to_string()
+        } else {
+            req.uri().path().to_string()
+        };
+
+        let span = if let Some(span) = &self.span {
+            let span = span.clone();
+            span.record("method", &display(method));
+            span.record("path", &display(path));
+            span
+        } else {
+            tracing::span!(
+                Level::INFO,
+                "http-request",
+                method = %method,
+                path = %path,
+                headers = tracing::field::Empty,
+            )
+        };
+
+        if self.record_headers {
+            span.record("headers", &debug(req.headers()));
+        }
+
+        let future = {
+            let _guard = span.enter();
+            tracing::info!(message = "received request");
+            self.inner.call(req)
+        };
+
+        ResponseFuture {
+            future,
+            span,
+            start,
+            latency_unit: self.latency_unit,
+            get_trace_status: self.get_trace_status.clone(),
+        }
+    }
+}
+
+#[pin_project]
+#[derive(Debug)]
+pub struct ResponseFuture<F, T> {
+    #[pin]
+    future: F,
+    span: Span,
+    start: Instant,
+    latency_unit: LatencyUnit,
+    get_trace_status: T,
+}
+
+impl<F, ResBody, E, T> Future for ResponseFuture<F, T>
+where
+    F: Future<Output = Result<Response<ResBody>, E>>,
+    T: GetTraceStatus<Response<ResBody>, E>,
+{
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+
+        let _guard = this.span.enter();
+
+        let result = ready!(this.future.poll(cx));
+        let time = this.start.elapsed();
+
+        match (
+            this.get_trace_status.trace_status(&result),
+            *this.latency_unit,
+        ) {
+            (TraceStatus::Status(status), LatencyUnit::Nanos) => {
+                tracing::info!(
+                    message = "completed request",
+                    status = status,
+                    latency_ns = %time.as_nanos(),
+                );
+            }
+            (TraceStatus::Status(status), LatencyUnit::Millis) => {
+                tracing::info!(
+                    message = "completed request",
+                    status = status,
+                    latency_ms = %time.as_millis(),
+                );
+            }
+            (TraceStatus::Error, LatencyUnit::Nanos) => {
+                tracing::info!(
+                    message = "request failed",
+                    latency_ns = %time.as_nanos(),
+                );
+            }
+            (TraceStatus::Error, LatencyUnit::Millis) => {
+                tracing::info!(
+                    message = "request failed",
+                    latency_ms = %time.as_millis(),
+                );
```

Comment:
this feels a bit opinionated about what the output looks like...do we want to consider providing some way of allowing the user to customize these events further?

Reply:
Yeah true. I guess you're probably the one to ask about this. I struggled with making things customizable since the name of span and the fields (afaik) must be declared up front.

I suppose we could do what we do wrt "response classification", have a trait for customizing it and provide a default impl. Off of the top of my head that trait would have a method like `fn emit_event(&self, trace_status: TraceStatus, latency_unit: LatencyUnit)`. `TraceStatus` would most likely be renamed though but thats separate.

What do you think about this?